### DISCUSSION TITLE
Supporting multiple catalogs

### DISCUSSION THREAD

**Discussion by roee88** [fybrik/fybrik#262] [Developers]
We currently support a single catalog connector, single credentials connector and up to two policy managers. As part of the work on Taxonomy and on Ingest scenario we decided to add support for multiple catalogs.

We don't have zero config yet, so the basic idea is to change the configmap to include this info. I had a couple of iterations of how to the configuration would look like without making it too complex yet keeping it relatively easy to implement the actual logic in manager and would appreciate feedback.

The structure I currently have looks like this (this is the most complex example that no one will ever need to use):
```yaml
catalogs:
  - url: http://wkc-connector-url
  - url: http://inline-catalog-connector-url
  - url: http://egeria-connector-url
    credentials: http://vault-connector-url
policy:
- url: http://opa-connector-url
```

Basically:
* Every item in `catalogs` :
    1. Must be a catalog connector 
    2. Must be a credentials connector to be used for the assets of that catalog (defaults to same connector but you can specify alternative like in the Egeria case)
    3. Optionally can be a policy manager connector if the connector implements this interface
* Items in `policy` are additional generic policy manager connectors not tied to a specific catalog.
* For the ingest scenario we discussed marking the "main" data lake catalog as well (not shown here but it should be)

Feedback?

**Comment by ronenkat** [comment]
couple of questions.
1. Is order important?
2. how do we map assets to catalogs, we only have a URL
3. how do we know if a policy manager is implemented (is it defined somewhere else?)
4. when do we use the generic policy, and when do we use the catalog policies

I would not use the term inline catalog, need a name that reflect its the sample/built-in M4D catalog if other option is not available

**Reply by roee88** [reply]
1. No
2. Good point, we can add a name for now but later when we stop using the weird JSON as asset IDs we can come up with a better alternative.
3. I thought about detect by just trying. I opted for this option to reduce configuration complexity because it felt fairly easy to implement it.
4. Similar to the current logic. We always query the PM related to the asset (when we have one). We add on top of that anything from the generic (policy compiler merging logic).

I'm using the name inline just for the sake of the discussion. In the PoC I'm working on I codenamed it Katalog.

**Reply by simanadler** [reply]
>     1. No
> 
>     2. Good point, we can add a name for now but later when we stop using the weird JSON as asset IDs we can come up with a better alternative.

I agree.  We need a name or an ID because every data set will now have to have a catalog ID and an asset ID.
> 
>     3. I thought about detect by just trying. I opted for this option to reduce configuration complexity because it felt fairly easy to implement it.

I thought that we had separated the data catalog functionality from the policy manager functionality.  That's why there are two separate connectors.  Right?  If there is any connection between them, I would think that the policy manager would indicate with which catalog(s) it knows how to work, rather than vice versa.
> 
>     4. Similar to the current logic. We always query the PM related to the asset (when we have one). We add on top of that anything from the generic (policy compiler merging logic).
> 
> 
> I'm using the name inline just for the sake of the discussion. In the PoC I'm working on I codenamed it Katalog.

**Reply by roee88** [reply]
I agree. Please provide a YAML suggestion. It helps get things clearer. I thought that we could just query them all and if the connector doesn't support what we ask it then it should return the equivalent of "don't ask me".

**Comment by simanadler** [comment]
When a data set is received it has a catalog id and a data set id, so M4DApplication controller knows from which catalog to read.  However, when registering a new data set in the enterprise catalog during ingest or a write flow ... how does M4DApplication controller know to which catalog to write?  There needs to be some indication in the catalog config doesn't there?

**Reply by roee88** [reply]
Catalog id is not the same. A single connector may interact with multiple catalogs of the same catalog system. That's why in the ingest scenario issue I suggested to make this explicit as part of the input.

**Reply by simanadler** [reply]
So if I understand correctly, the user passes the catalog name and asset ID?  M4DApplication controller then has to iterate over each of the catalog connectors listed in the config file and ask each connector if it supports the declared catalog?  

If this is correct, then the catalog names will be in the taxonomy I assume.  Right?

**Reply by roee88** [reply]
That's not what I meant. I meant to extend the input to include the catalog system name. The equivalent of egeria/coco-catalog/myasset.

**Reply by simanadler** [reply]
OK.  It can be done that way.  In your example though "egaria" would still have to come from the taxonomy.  Right?

**Reply by roee88** [reply]
It has to match the name set for the connector. Not sure if it's a taxonomy thing or just hardcoded in the connector (i.e. well known name publicized by the connector).

**Reply by simanadler** [reply]
The user filling out egeria/coco-catalog/myasset in the M4DApplication yaml will not know what connector is deployed in the M4D control plane nor what it is called.  He will only know that he used ABCCatalog, Egaria, or XYZCatalog to choose the data.  Until now we assumed there is only one enterprise catalog, so knowing the name was sufficient.  What happens when there is more than one instance of ABCCatalog deployed in the enterprise?  How do we know which instance he means?  How do we know our connector is connected to the correct instance?  When / how do we decide to use each catalog and for what?

Personally I'm not sure whether we should be trying to address all of these questions.  I thought that for MVP we would assume a single enterprise catalog.  If we want a second to address only external data that makes sense to me.  I.e. A temporary way to handle data being ingested or written outside of the data lake.

In such a catalog though, who owns the data lifecycle?  If source data is registered in the Katalog (inline catalog) when ingest is initiated, for example, should it be deleted from the Katalog after the ingest is completed?

**Reply by roee88** [reply]
> Personally I'm not sure whether we should be trying to address all of these questions. I thought that for MVP we would assume a single enterprise catalog. If we want a second to address only external data that makes sense to me. I.e. A temporary way to handle data being ingested or written outside of the data lake.

If it's just for ingest then we don't have to get it in the project at this point. Use a catalog system that supports multiple catalogs for a demonstration of ingest. 

> In such a catalog though, who owns the data lifecycle? If source data is registered in the Katalog (inline catalog) when ingest is initiated, for example, should it be deleted from the Katalog after the ingest is completed?

There is no special catalog for ingest. We don't special case ingest or anything similar. It's a catalog like any other catalog.

**Reply by roee88** [reply]
The hierarchy I described is `<system>/<fully qualified asset identifier>`. In systems like WKC the fully qualified asset identifier represents a catalog id and and an asset within it. In others like Azure it's a single catalog per organization. Your description is of a single organization using multiple instances of the same `<system>` - is that a real thing?

**Comment by simanadler** [comment]
1) For the ingest scenario, am I correct in assuming that there will be two catalogs in use?  (1) In which the data owner provides the information and credentials for the source data?  (2) the enterprise catalog into which the destination data set is registered?

2) What are the other scenarios in which you envision the need for multiple catalogs?

**Reply by roee88** [reply]
I just have multiple catalog systems that I work with because the set if features they provide solve different use cases within my organization (this isn't something that I made up)

**Reply by simanadler** [reply]
Can you share some concrete examples?  I'm not sure I'm following.

**Comment by simanadler** [comment]
@roee88 To summarize what we discussed by phone earlier today....

**The problems we are trying to address** by creating an inline catalog service and by supporting multiple catalog services are:
1) Provide a built in, simple catalog that can be used as part of a quick start configuration for M4D
2) Provide a mechanism for the data owner / data user to provide information about data sets outside the controlled environment, i.e. data sets not registered in the cataloged environment.  Examples of such situations are:
* When data is being initially loaded into the environment, the data owner provides the endpoint information of the source data set
* (future) When writing data to a data store outside the controlled environment.  Ex: export, archive

To support #2 means that M4D must support two catalog services in parallel.  Thus, we have a need for supporting multiple catalogs.  

**Out of scope:** If an enterprise has multiple data catalog services, M4D interfaces with only one.  If synchronization between them is required, there are solutions such as egaria, which is out of scope for M4D.

**Assumption**: M4D will support connectors to different policy engines, but will connect to only one in any given M4D deployment.

**Comment by roee88** [comment]
Updated suggestion for a next step:

```yaml
connectors:
  catalog:
  - name: connector-service-name
    default: true ## TODO: do we want this field?
  - name: another-connector-service-name
  - name: catalog-only-connector-service-name
    credentials: credentials-connector-service-name
  policy: 
  - name: policy-manager-connector-service-name
```

1. `default` is for making the catalog field optional in `M4DApplication`. We can alternatively decide that it's always required. If we decide to have the `default` keyword then the configuration must be validated (i.e., one and only one is set as default). 
2. `credentials` is only required if credentials are served by a different connector. 
3. Policy connectors is a list of up to two items in this iteration. Next step would be to make it a single item if consensus is reached about that. 

Can be a configmap or a configuration CRD, but the configuration should not be used as environment variables and changes to it should trigger a change in manager.


@simanadler @rohithdv opinions ?

**Reply by simanadler** [reply]
@roee88 What component should check that what is declared here matches what is deployed in the control plane? Ex: If it is written here that the inline catalog connector is deployed but in reality only the wkc connector is deployed, when/how should this error be noted?  I would hope that it would happen during the control plane deployment.  Right?

**Reply by roee88** [reply]
For now it would just fail to connect. Eventually as much of it would be part of the deployment but we can't fully validate as part of deployment because connectors are third parties.

**Reply by shlomitk1** [reply]
> ```yaml
> connectors:
>   catalog:
>   - name: connector-service-name
>     default: true ## TODO: do we want this field?
>   - name: another-connector-service-name
>   - name: catalog-only-connector-service-name
>     credentials: credentials-connector-service-name
>   policy: 
>   - name: policy-manager-connector-service-name
> ```
> 
Instead of a boolean default I suggest to use default as a service name pointing to a specified one, for example:
```
catalog:
  - catalog1: http://catalog1:3000
     credentials:  http://creds:3002
  - catalog2: http://catalog2:4000
  - default: http://catalog1:3000
     credentials:  http://creds:3002
```

**Reply by roee88** [reply]
Would you create two clients for catalog1 then?

**Reply by shlomitk1** [reply]
> 
> 
> Would you create two clients for catalog1 then?

No. Unless I'm missing something, I thought this is a map of services replacing what we have today in the config map. Today, we search for CATALOG_CONNECTOR_URL or similar to find the required service. With the multiple deployments, we could either select catalog1, catalog2 or both. If a default is catalog1 but it is not specified, a default service will be connected (catalog1).

**Reply by roee88** [reply]
I'm confused about listing http://catalog1:3000 twice in your example. It looks like another catalog that is named `default` that happens to have the same values as a different entry. 

Anyway, the idea of the structure I posted was to use the service name as the meaningful identifying name of the connector and avoid using full URLs. The requirement is that a connector `Service` will have a meaningful name and use `port: 80` (while `targetPort` is different). For example, with the following config the manager holds a mapping from "katalog-connector" to a grpc client connected to http://katalog-connector:80 and from "egeria-connector" to a grpc client connected to http://egeria-connector:80. When an M4DApplication specifies `catalogService: egeria-connector` we use the client from the mapping (we can prune the "-connector" suffix):

```yaml
connectors:
  catalog:
  - name: katalog-connector
  - name: egeria-connector
    credentials: vault-connector
```

If no `catalogService` is specified in `M4DApplication` then we need to use a default. I'm not sure how to adopt your suggestion here as it changes all of the above (but it's a valid complete alternative)

**Comment by simanadler** [comment]
In the meeting earlier today @roee88 you mentioned a concern about using multiple catalogs in parallel, as was proposed earlier for the ingest scenario.  I believe your concern was that different catalogs would require different parameters when reading/registering an asset in the catalog, and that we don't want to find ourselves creating a connector that knows how to work with all of them - i.e. reinventing egaria.

Instead, you suggested that we use the same catalog service for both steps of the ingest.
1) User provides the endpoint to the source data and relevant metadata in a catalog belonging to the user
2) M4D reads the info into the user catalog, copies the data from the transactional system into the data lake, and then registers the copy in **enterprise** catalog

Is this a correct understanding @roee88 ?  If so, 
* we would have two different sets of credentials for the catalog service.  Right?  One for the ingest user, and one for registering in the enterprise catalog.
* Do we still need the config described above if we are only using one catalog service at a time?
* How do we know which is the "enterprise catalog" to which we register the new data set?

@rohithdv @shlomitk1 @ronenkat Your thoughts?

**Reply by roee88** [reply]
Giving my answers but please see my [message](https://github.com/IBM/the-mesh-for-data/discussions/262#discussioncomment-309620)  below

> we would have two different sets of credentials for the catalog service. Right? One for the ingest user, and one for registering in the enterprise catalog.

Pretty sure that it should be the same but we can verify if it makes sense if we ever reach that point.

> Do we still need the config described above if we are only using one catalog service at a time?

No

> How do we know which is the "enterprise catalog" to which we register the new data set?

The target catalog id is already explicit in the spec.

**Comment by roee88** [comment]
@simanadler and all,

Indeed, I have raised the concern you described. I can answer your questions above about the suggestion I made during the call but I don't think that with our current design it's a good suggestion actually, for reasons unrelated to your questions. The asset to catalog is actually the copy we made which according to the current design of storage provisioning has nothing to do with the original asset or the catalog system. It will also not help us in write scenarios that create a new asset. Typically, everything that we provision on our own, unless we tie it to the catalog connector somehow, will be an issue. So I think that it's either to create a common dominator for the storage types that we use for write and copies, or to tie the provisioning to the catalog connector in some way.

I need to think about it more. It's similar to the problem I encountered in the work on taxonomies where the manager didn't care about how connections look like, except for S3 because of provisioning.

**Reply by simanadler** [reply]
@roee88 OK.  What are the next steps for your perspective?  We need to implement something in the next week or so.

**Reply by roee88** [reply]
What can I say? Do what you have to do.. just consider it highly experimental so don't over invest.

**Reply by simanadler** [reply]
I would suggest that we adopt your suggestion @roee88 and use only one catalog at any given time.  When ingesting new data from an external source the data owner will put the info about the endpoint and the asset's metadata in catalogservice/usercatalog and M4D will register the new data set in catalogservice/enterprisecatalog after it has been copied.  Both catalog details are specified in M4DApplication.  

There will be no need for multiple catalogs in a single M4D deployment (and thus no need for the config discussed previously).  However, different deployments may use different catalogs.  Catalogs currently supported are: CRD based catalog created by @roee88, egaria, Watson Knowledge Catalog.

@roee88 @shlomitk1 @rohithdv @ronenkat  Please counter or confirm this proposal by end of day tomorrow (Wed).  Thanks!

**Reply by roee88** [reply]
Fine by me but I didn't understand how that solves the registering of a copy.

**Reply by simanadler** [reply]
Need to add an API to the catalog connectors for registering a data asset.  @roee88 @shlomitk1 @rohithdv

**Reply by roee88** [reply]
I meant that this new API will need to accept connection information and since it's of a copy then it's connection information using structure that we define and hence we get back to the problem I described although we use a single catalog system (i.e., it doesn't help at all). I'm fine with designing such an API as something experimental and later attack this issue.

**Reply by simanadler** [reply]
@roee88 Yes, the manager / catalog connector API for registering a new data asset in the catalog will need to be generic enough to support the different catalogs to the best of our ability.  Each specific catalog connector will need to then call the specific catalog service APIs for registering the asset.  I don't see any way to get around this at the moment.  @rohithdv Input from you and your team would be greatly appreciated.

**Reply by rohithdv** [reply]
Yes I think this is the best way forward @simanadler @roee88. My suggestion is the following. Let us now focus on one specific catalog and get it working (Since we have already invested time on WKC, WKC would be my choice for this). Later on we can add support for other catalogs like the inline catalog, Egeria, AWS, Amundsen etc. I think it is a very  difficult to support all catalogs at once. We should do it incrementally. 

We have already started working on the WKC registration of an asset feature now. I think we should design the protos for registration of an asset keeping WKC in mind for now as supporting for WKC itself is a non trivial task in my opinion. This will help us to simplify the complex task at hand. Once WKC is fully supported then we can move to other catalogs incrementally modifying the protos wherever necessary ( without breaking the existing catalog support ). Please let me know your views on this. Thanks.
